File: python_implementation/discretize.py
import math

class Discretizador:
    # Diccionario para almacenar puntos de corte por método de discretización
    puntos_corte = {}

    @staticmethod
    def ancho_igual(columna: list[float | int], bins: int) -> list[int]:
        '''
        Método para discretizar una columna en intervalos de ancho igual.
        Parameters:
            columna: (list[float | int]) Lista de valores numéricos a discretizar.
            bins: (int) Número de intervalos (bins).
        Return:
            list[int]: Lista de valores discretizados en función del intervalo correspondiente.
        '''
        min_valor, max_valor = min(columna), max(columna)
        rango = max_valor - min_valor + 1e-10  # Para evitar divisiones por cero
        ancho = rango / bins
        puntos_corte = [min_valor + i * ancho for i in range(1, bins)]
        Discretizador.puntos_corte['ancho_igual'] = puntos_corte
        discretizado = [
            min(bins - 1, int((valor - min_valor) / rango * bins)) for valor in columna
        ]
        return discretizado

    @staticmethod
    def frecuencia_igual(columna: list[float | int], bins: int) -> list[int]:
        '''
        Método para discretizar una columna en intervalos de frecuencia igual.
        Parameters:
            columna: (list[float | int]) Lista de valores numéricos a discretizar.
            bins: (int) Número de intervalos (bins).
        Return:
            list[int]: Lista de valores discretizados en función del intervalo correspondiente.
        '''
        valores_ordenados = sorted(columna)
        n = len(columna)
        tamano_bin = n // bins
        puntos_corte = [valores_ordenados[(i + 1) * tamano_bin - 1] for i in range(bins - 1)]
        Discretizador.puntos_corte['frecuencia_igual'] = puntos_corte
        discretizado = [0] * n
        for indice_bin in range(bins):
            inicio = indice_bin * tamano_bin
            fin = inicio + tamano_bin if indice_bin < bins - 1 else n
            for i in range(inicio, fin):
                discretizado[columna.index(valores_ordenados[i])] = indice_bin
        return discretizado

    @staticmethod
    def k_means(columna: list[float | int], bins: int) -> list[int]:
        '''
        Método para discretizar una columna usando el algoritmo de k-means.
        Parameters:
            columna: (list[float | int]) Lista de valores numéricos a discretizar.
            bins: (int) Número de intervalos (bins).
        Return:
            list[int]: Lista de valores discretizados en función del intervalo correspondiente.
        '''
        min_valor, max_valor = min(columna), max(columna)
        centroides = [min_valor + (i + 0.5) * (max_valor - min_valor) / bins for i in range(bins)]
        centroides_previos = [None] * bins
        while centroides != centroides_previos:
            clusters = [[] for _ in range(bins)]
            for valor in columna:
                indice_mas_cercano = min(range(bins), key=lambda i: abs(valor - centroides[i]))
                clusters[indice_mas_cercano].append(valor)
            centroides_previos = centroides[:]
            centroides = [sum(cluster) / len(cluster) if cluster else centroides[i]
                          for i, cluster in enumerate(clusters)]
        Discretizador.puntos_corte['k_means'] = [(centroides[i] + centroides[i + 1]) / 2 for i in range(bins - 1)]
        discretizado = [
            next((i for i, corte in enumerate(Discretizador.puntos_corte['k_means']) if valor < corte), bins - 1)
            for valor in columna
        ]
        return discretizado

    @staticmethod
    def basado_en_cuantiles(columna: list[float | int], bins: int) -> list[int]:
        '''
        Método para discretizar una columna basada en cuantiles.
        Parameters:
            columna: (list[float | int]) Lista de valores numéricos a discretizar.
            bins: (int) Número de intervalos (bins).
        Return:
            list[int]: Lista de valores discretizados en función del intervalo correspondiente.
        '''
        valores_ordenados = sorted(columna)
        n = len(valores_ordenados)
        tamano_bin = n / bins
        puntos_corte = [(valores_ordenados[int(i * tamano_bin - 1)] + valores_ordenados[int(i * tamano_bin)]) / 2 for i in range(1, bins)]
        Discretizador.puntos_corte['basado_en_cuantiles'] = puntos_corte
        return [next((i for i, corte in enumerate(puntos_corte) if valor < corte), bins - 1) for valor in columna]

    @staticmethod
    def discretizar_columna(df, nombre_columna: str, metodo: str = 'ancho_igual', bins: int = 3) -> None:
        '''
        Método para discretizar una columna de un data.frame usando un método específico.
        Parameters:
            df: (objeto) DataFrame con la columna a discretizar.
            nombre_columna: (str) Nombre de la columna a discretizar.
            metodo: (str) Método de discretización ('ancho_igual', 'frecuencia_igual', 'k_means', 'basado_en_cuantiles').
            bins: (int) Número de intervalos (bins).
        Return:
            None
        '''
        columna = df.obtener_columna(nombre_columna)
        if metodo == 'ancho_igual':
            discretizado = Discretizador.ancho_igual(columna, bins)
        elif metodo == 'frecuencia_igual':
            discretizado = Discretizador.frecuencia_igual(columna, bins)
        elif metodo == 'k_means':
            discretizado = Discretizador.k_means(columna, bins)
        elif metodo == 'basado_en_cuantiles':
            discretizado = Discretizador.basado_en_cuantiles(columna, bins)
        else:
            raise ValueError("Método de discretización no válido. Use 'ancho_igual', 'frecuencia_igual', 'k_means', o 'basado_en_cuantiles'.")
        df.datos[nombre_columna] = discretizado

    @staticmethod
    def obtener_rangos_bins(metodo: str = 'ancho_igual', decimales: int = 8) -> list[tuple[float, float]]:
        '''
        Método para obtener los rangos de cada bin en función del método de discretización usado.
        Parameters:
            metodo: (str) Método de discretización para el cual obtener los rangos de bins.
            decimales: (int) Número de decimales para redondear los puntos de corte.
        Return:
            list[tuple[float, float]]: Lista de tuplas que representan los rangos de cada bin.
        Exceptions:
            ValueError si no existen puntos de corte para el método especificado.
        '''
        if metodo not in Discretizador.puntos_corte:
            raise ValueError("No existen puntos de corte para el método especificado.")
        
        puntos_corte = Discretizador.puntos_corte[metodo]
        rangos_bins = []
        corte_previo = float('-inf')
        for corte in puntos_corte:
            rangos_bins.append((corte_previo, round(corte, decimales)))
            corte_previo = round(corte, decimales)
        rangos_bins.append((corte_previo, float('inf')))
        
        return rangos_bins

File: R_implementation/discretize.R
# Define Discretizador as an environment-based S3 class
Discretizador <- function() {
  e <- new.env()
  e$puntos_corte <- list()
  class(e) <- "Discretizador"
  return(e)
}

#' Método para discretizar una columna en intervalos de ancho igual.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param columna (numeric) Lista de valores numéricos a discretizar.
#' @param bins (int) Número de intervalos (bins).
#' @return (numeric) Lista de valores discretizados en función del intervalo correspondiente.
ancho_igual <- function(discretizador, columna, bins) {
  min_valor <- min(columna)
  max_valor <- max(columna)
  rango <- max_valor - min_valor + 1e-10
  ancho <- rango / bins
  puntos_corte <- min_valor + seq_len(bins - 1) * ancho
  discretizador$puntos_corte[["ancho_igual"]] <- puntos_corte
  discretizado <- sapply(columna, function(valor) {
    min(bins - 1, floor((valor - min_valor) / rango * bins))
  })
  return(discretizado)
}

#' Método para discretizar una columna en intervalos de frecuencia igual.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param columna (numeric) Lista de valores numéricos a discretizar.
#' @param bins (int) Número de intervalos (bins).
#' @return (numeric) Lista de valores discretizados en función del intervalo correspondiente.
frecuencia_igual <- function(discretizador, columna, bins) {
  valores_ordenados <- sort(columna)
  n <- length(columna)
  tamano_bin <- n %/% bins
  puntos_corte <- valores_ordenados[(1:(bins - 1)) * tamano_bin]
  discretizador$puntos_corte[["frecuencia_igual"]] <- puntos_corte
  discretizado <- numeric(n)
  for (indice_bin in 0:(bins - 1)) {
    inicio <- indice_bin * tamano_bin + 1
    fin <- if (indice_bin < bins - 1) inicio + tamano_bin - 1 else n
    indices <- match(valores_ordenados[inicio:fin], columna)
    discretizado[indices] <- indice_bin
  }
  return(discretizado)
}

#' Método para discretizar una columna usando el algoritmo de k-means.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param columna (numeric) Lista de valores numéricos a discretizar.
#' @param bins (int) Número de intervalos (bins).
#' @return (numeric) Lista de valores discretizados en función del intervalo correspondiente.
k_means <- function(discretizador, columna, bins) {
  min_valor <- min(columna)
  max_valor <- max(columna)
  centroides <- min_valor + ((0:(bins - 1)) + 0.5) * (max_valor - min_valor) / bins
  centroides_previos <- rep(NA, bins)
  
  while (!all(centroides == centroides_previos, na.rm = TRUE)) {
    clusters <- vector("list", bins)
    for (valor in columna) {
      distancias <- abs(valor - centroides)
      indice_mas_cercano <- which.min(distancias)
      clusters[[indice_mas_cercano]] <- c(clusters[[indice_mas_cercano]], valor)
    }
    centroides_previos <- centroides
    centroides <- sapply(seq_len(bins), function(i) {
      if (length(clusters[[i]]) > 0) {
        mean(clusters[[i]])
      } else {
        centroides[i]
      }
    })
  }
  
  puntos_corte <- (centroides[-bins] + centroides[-1]) / 2
  discretizador$puntos_corte[["k_means"]] <- round(puntos_corte, 8)
  discretizado <- sapply(columna, function(valor) {
    idx <- which(valor < puntos_corte)[1]
    if (is.na(idx)) bins - 1 else idx - 1
  })
  return(discretizado)
}

#' Método para discretizar una columna basada en cuantiles.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param columna (numeric) Lista de valores numéricos a discretizar.
#' @param bins (int) Número de intervalos (bins).
#' @return (numeric) Lista de valores discretizados en función del intervalo correspondiente.
basado_en_cuantiles <- function(discretizador, columna, bins) {
  probabilidades <- seq(0, 1, length.out = bins + 1)
  bordes_bin <- quantile(columna, probs = probabilidades)
  puntos_corte <- bordes_bin[-c(1, bins + 1)]
  discretizador$puntos_corte[["basado_en_cuantiles"]] <- round(puntos_corte, 8)
  
  discretizado <- cut(columna, breaks = bordes_bin, labels = FALSE, include.lowest = TRUE) - 1
  return(discretizado)
}

#' Método para discretizar una columna de un data.frame usando un método específico.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param df (data.frame) DataFrame con la columna a discretizar.
#' @param nombre_columna (str) Nombre de la columna a discretizar.
#' @param metodo (str) Método de discretización ('ancho_igual', 'frecuencia_igual', 'k_means', 'basado_en_cuantiles').
#' @param bins (int) Número de intervalos (bins).
#' @return (data.frame) El data.frame con la columna discretizada.
discretizar_columna <- function(discretizador, df, nombre_columna, metodo = "ancho_igual", bins = 3) {
  columna <- obtener_columna(df, nombre_columna)
  if (metodo == "ancho_igual") {
    discretizado <- ancho_igual(discretizador, columna, bins)
  } else if (metodo == "frecuencia_igual") {
    discretizado <- frecuencia_igual(discretizador, columna, bins)
  } else if (metodo == "k_means") {
    discretizado <- k_means(discretizador, columna, bins)
  } else if (metodo == "basado_en_cuantiles") {
    discretizado <- basado_en_cuantiles(discretizador, columna, bins)
  } else {
    stop("Método de discretización no válido. Use 'ancho_igual', 'frecuencia_igual', 'k_means', o 'basado_en_cuantiles'.")
  }
  
  df$datos[[nombre_columna]] <- discretizado
  return(df)
}

#' Método para obtener los rangos de cada bin en función del método de discretización usado.
#'
#' @param discretizador (Discretizador) Objeto de la clase Discretizador.
#' @param metodo (str) Método de discretización para el cual obtener los rangos de bins.
#' @param decimales (int) Número de decimales para redondear los puntos de corte.
#' @return (list) Lista de tuplas que representan los rangos de cada bin.
#' @examples
#' # Ejemplo de uso
#' discretizador <- Discretizador()
#' columna <- c(1, 2, 3, 4, 5)
#' discretizado <- ancho_igual(discretizador, columna, bins = 3)
#' rangos <- obtener_rangos_bins(discretizador, metodo = "ancho_igual")
obtener_rangos_bins <- function(discretizador, metodo = "ancho_igual", decimales = 8) {
  if (!(metodo %in% names(discretizador$puntos_corte))) {
    stop("No existen puntos de corte para el método especificado.")
  }
  puntos_corte <- discretizador$puntos_corte[[metodo]]
  rangos_bins <- list()
  corte_previo <- -Inf
  for (corte in puntos_corte) {
    rangos_bins <- append(rangos_bins, list(c(corte_previo, round(corte, decimales))))
    corte_previo <- round(corte, decimales)
  }
  rangos_bins <- append(rangos_bins, list(c(corte_previo, Inf)))
  return(rangos_bins)
}
